streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java

```diff
@@ -160,11 +160,18 @@ public void processInOrder(final K key, final V value, final long timestamp) {
 
                     if (endTime < timestamp) {
                         leftWinAgg = next.value;
+                        // store the combined window if it is found so that a right window can be created for
+                        // the combined window's max record, as needed
                         if (isLeftWindow(next) || endTime == windows.timeDifferenceMs()) {
                             latestLeftTypeWindow = next;
                         }
                     } else if (endTime == timestamp) {
                         leftWinAlreadyCreated = true;
+                        // if current record's left window is the combined window, need to check later if there is a
+                        // record that needs a right window within the combined window
+                        if (endTime == windows.timeDifferenceMs()) {
+                            latestLeftTypeWindow = next;
+                        }
```

Comment:
I see what you're saying about checking to see if the left window (in this case [0,timeDifference]) was already a left window and I can substitute for that check. The overall point was to catch instances of a record coming in at say 10 (with a timeDifference of 10) and also having one at 5, so 5 needs a right window but the combined window will only be picked up by this `if()` since `endTime == timestamp`. Setting `latestLeftTypeWindow = next` here will generally _not_ be a left type window, but the combined window is a special case since there are records that may need a right window in there anyways. Because this tracks the _latest_ left type window, storing the combined window should never override any other formal left type window that needs a right window.

That being said, I get that this is confusing. Do you think changing the check to `if (endTime == windows.TimeDifferenceMs() && !isLeftWindow(next)) would make it seem cleaner?

Reply:
> That being said, I get that this is confusing. Do you think changing the check to `if (endTime == windows.TimeDifferenceMs() && !isLeftWindow(next)) would make it seem cleaner?
 Haha no, I don't think saying `if (!isLeftWindow(next)): then next = latestLeftTypeWindow` would be less confusing. If we call a variable `leftTypeWindow` then it should _always_ be a left type window.

That said, I now see what you meant here and it's the same problem as above, with the same fix of replacing `latestLeftTypeWindow` with `previousRecordTimestamp`. In that case I think we can just remove this check entirely (ie, don't explicitly check if it's the combined window), and all we need to do is make sure `previousRecordTimestamp` is set correctly

Reply:
I think with this replacement then we might be able to get out of doing any kind of special handling for the combined window outside of `processEarly`

Reply:
I think we'd still need a check within this part of the `if()` to catch the previous timestamp, right? I think we could take out the `if (endTime == windows.TimeDifferenceMS())` and instead always do `previousRecord = next.value.timestamp`

Reply:
Yeah sorry I didn't mean that we shouldn't have any conditionals here whatsoever, I just meant that we don't need the combined window check (or really anything other than what we need to accurately set `previousRecordTimestamp`)

Reply:
I guess you could just have a separate `setPreviousRecordTimestampIfNecessary(window, previousRecordTimestamp)` method that sets the `previousRecordTimestamp` to the window's max timestamp if it's larger. And then if it ends up that `previousRecordTimestamp == timestamp` then we can automatically skip all of the window creation

Reply:
Would the window's max timestamp always be larger than the previous window's by nature of moving forward in time? I think it would be somewhat more convenient but I feel like adding additional boolean checks makes the algorithm feel longer, especially if the issue will get handled in later checks